docs/sources/developers/plugins/add-authentication-for-data-source-plugins.md

```diff
@@ -148,25 +230,59 @@ To set the body content and length of proxied requests, use the `body` property.
 ]
 ```
 
-### Enable token authentication
+### Add a OAuth 2.0 proxy route to your plugin
+
+The data source proxy supports OAuth 2.0 authentication.
 
-To enable token-based authentication for proxied requests, use the `tokenAuth` property.
+Since the request to each route is made server-side, only machine-to-machine authentication is supported. In order words, if you need to use a different grant than client credentials, you need to implement it yourself.
 
-Grafana automatically renews the token when it expires.
+To authenticate using OAuth 2.0, add a `tokenAuth` object to the proxy route definition. If necessary, Grafana performs a request to the URL defined in `tokenAuth` to retrieve a token before making the request to the URL in your proxy route. Grafana automatically renews the token when it expires.
+
+Any parameters defined in `tokenAuth.params` are encoded as `application/x-www-form-urlencoded` and sent to the token URL.
 
 ```json
-"routes": [
-  {
-    "path": "example",
-    "url": "https://api.example.com",
-    "tokenAuth": {
-      "url": "https://login.example.com/oauth2/token",
-      "params": {
-        "grant_type": "client_credentials",
-        "client_id": "{{ .JsonData.clientId }}",
-        "client_secret": "{{ .SecureJsonData.clientSecret }}"
+{
+  "routes": [
+    {
+      "path": "api",
+      "url": "https://api.example.com/v1",
+      "tokenAuth": {
+        "url": "https://api.example.com/v1/oauth/token",
+        "params": {
+          "grant_type": "client_credentials",
+          "client_id": "{{ .SecureJsonData.clientId }}",
+          "client_secret": "{{ .SecureJsonData.clientSecret }}"
+        }
       }
     }
+  ]
+}
+```
+
+## Authenticate using a backend plugin
+
+While the data source proxy supports the most common authentication methods for HTTP APIs, using proxy routes has a few limitations:
+
+- Proxy routes only support HTTP
+- Proxy routes don't support custom token authentication
+
+If any of these limitations apply to your plugin, you need to add a [backend plugin](TODO). Since backend plugins run on the server they can access decrypted secrets, which makes it easier to implement custom authentication methods.
+
+The decrypted secrets are available from the `DecryptedSecureJSONData` field in the instance settings.
+
+```go
+func (ds *dataSource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+  instance, err := ds.im.Get(req.PluginContext)
```

Comment:
This example uses the instance manager (im) but might be a bit unclear/confusing leaving that out in the example? 

This could have been nice to link to if it would have included usage of jsonData/secureJsonData

Reply:
I struggled to make an example for this that showcases the DecryptedSecureJsonData without having to include the bigger part of the plugin for context. What would be minimum amount of context we'd need to provide to demonstrate the specific use case of using the secure JSON data?

Reply:
It's tricky. So perhaps just something like this could work?

```go
func (ds *dataSource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
  if mySecret, exists := req.PluginContext.DataSourceInstanceSettings.DecryptedSecureJSONData["mySecret"]; exists {
    // Call my custom authentication method using mySecret value...
  }
```